sem: consolidate macro pragma processing

zerbina opened the pull request:
## Summary

Merge the separate implementations of macro/template pragma processing
into a single one. The benefits:
* less duplicated code
* macro and template annotations are ensured to work the same for all
  types of definitions
* a single place to change the behaviour in

This changes:
* routine AST passed to macro/template pragmas keeps the `nnkPragma`
  node, even if empty
* ill-formed accent quoted expressions preceding a macro/template
  pragma don't prevent the transform from taking place

## Details

Routines, types, and var/let/const definitions all had their own
implementation, but their only difference was in where the AST was
modified and whether a full copy was created or not.

There's nothing that requires fundamentally different processing for
each type of definition, so the processing is consolidated:
* the new `tryMacroPragma` procedure processes a single pragma,
  returning either the raw transform or nil
* the new `semAnnotation` processes a pragma list (`nkPragma`), trying
  each item with `tryMacroPragma`
* sem'ing of routines, types, and var/let/const definitions use
  `semAnnotation`

So that the macro/template pragma call can be temporarily removed from
the definition's pragma list (which is at different tree positions for
each definition AST), the address of the node is passed to
`semAnnotation`.

Differences compared to the previous implementation:
* empty `nkPragma` nodes are kept. This was previously only the case
  for var/let/const annotations
* instead of the manual `semMacroExpr` call, `afterCallActions` is
  used. This is a step towards removing the bare `semMacroExpr` calls

Ideally, `semAnnotation` would always return the raw transformation
result and leave sem'ing it to the callsite, but this is not yet
possible.

zerbina committed:
sem: consolidate macro pragma processing

Summary
=======

Merge the separate implementations of macro/template pragma processing
into a single one. The benefits:
* less duplicated code
* macro and template annotations are ensured to work the same for all
  types of definitions
* a single place to change the behaviour in

This changes:
* routine AST passed to macro/template pragmas keeps the `nnkPragma`
  node, even if empty
* ill-formed accent quoted expressions preceding a macro/template
  pragma don't prevent the transform from taking place

Details
=======

Routines, types, and var/let/const definitions all had their own
implementation, but their only difference was in where the AST was
modified and whether a full copy was created or not.

There's nothing that requires fundamentally different processing for
each type of definition, so the processing is consolidated:
* the new `tryMacroPragma` procedure processes a single pragma,
  returning either the raw transform or nil
* the new `semAnnotation` processes a pragma list (`nkPragma`), trying
  each item with `tryMacroPragma`
* sem'ing of routines, types, and var/let/const definitions use
  `semAnnotation`

So that the macro/template pragma call can be temporarily removed from
the definition's pragma list (which is at different tree positions for
each definition AST), the address of the node is passed to
`semAnnotation`.

Differences compared to the previous implementation:
* `considerQuotedIdent` errors are ignored. If not handled by a macro
  pragma, the `nkAccQuoted` will result in an error during the normal
  pragma processing
* empty `nkPragma` nodes are kept. This was previously only the case
  for var/let/const annotations
* instead of the manual `semMacroExpr` call, `afterCallActions` is
  used. This is a step towards removing the bare `semMacroExpr` calls

Ideally, `semAnnotation` would always return the raw transformation
result and leave sem'ing it to the callsite, but this is not yet
possible.

saem commented on compiler/sem/semstmts.nim:
```diff
@@ -480,6 +480,90 @@ proc copyExcept(n: PNode, i: int): PNode {.inline.} =
+    # a custom pragma as opposed to a built-in
+    let (ident, err) = considerQuotedIdent(c, key)
+    if err != nil:
+      # don't report the error yet, a macro/template pragma could
+      # still make the AST well-formed
```
This would then be an ordering violation? where any future macro pragma(s) that make it valid, if it exists, should have come first?

The biggest drawback I can think of is a spelling mistake in one pragma might go unreported because a second pragma clobbered it, leading to unexpected semantics.

The benefit of reporting the error, is that we can assume all pragma prior to the current `i` to have been fully applied, simplifying the concept of progress.

zerbina replied:
I'm not sure I understand the question with the ordering violation, but with the spelling mistake part I do somewhat agree.

Other invalid AST in pragma lists is already ignored during the early annotation processing, so my thinking was that making an exception for invalid accent quotations is unexpected/inconsistent.

zerbina replied:
Your edit helped, thanks. Hm, yeah, I see what you mean. The perspective I made the change from is that macro pragmas take precedence over anything else, operating on the unprocessed definition AST.

saem replied:
Sorry it wasn't clear from the get go, I was more tired than I realized and ended up with some word salad. Glad to hear the edit helped clarify things.

saem reviewed (commented):
The non-reporting of identifier errors might be quite problematic, see inline comment for more details. Happy to discuss it further, in case I'm misunderstanding something.

zerbina committed:
semstmts: revert the identifier error handling

Until it's clearer what macro/template pragmas should allow for and
what not, keep the previous behaviour of erroneous accent quotations
coming before any macro pragma resulting in an error.

saem commented:
/merge

saem commented:
/cancel

Just realized the PR body wasn't updated just yet.

zerbina commented:
> Just realized the PR body wasn't updated just yet.

Yep, sorry, I got sidetracked with something else. It's updated now.

zerbina commented:
/merge
